core/src/main/scala/com/softwaremill/react/kafka_api_opt1/Consumer.scala

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2014 - 2016 Softwaremill <http://softwaremill.com>
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package com.softwaremill.react.kafka_api_opt1
+
+import scala.collection.immutable
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+
+import akka.NotUsed
+import akka.stream.Graph
+import akka.stream.Inlet
+import akka.stream.Outlet
+import akka.stream.Shape
+import akka.stream.SourceShape
+import akka.stream.scaladsl.Flow
+import akka.stream.scaladsl.GraphDSL
+import akka.stream.scaladsl.Source
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import org.apache.kafka.clients.consumer.KafkaConsumer
+import org.apache.kafka.clients.consumer.OffsetAndMetadata
+import org.apache.kafka.common.TopicPartition
+
+object Consumer {
+
+  // FIXME those should probably be real case classes (or value classes)?
+  type CommitMsg = Map[TopicPartition, OffsetAndMetadata]
+  type CommitConfirmation = Future[CommitMsg]
+
+  case class ConsumerShape[K, V](
+      commit: Inlet[CommitMsg],
+      messages: Outlet[ConsumerRecord[K, V]],
+      confirmation: Outlet[CommitConfirmation]
```

Comment:
> The at-least-once example you show only works when dealing with a single action that returns a Future,

How so?

>  in particular it would be awkward to connect this kind of stream to a KafkaSink. 

What I proposed in another discussion here that KafkaSink takes an Acknowledgable element, not a raw element and it just calls `commit()` itself.

> Of course this is probably not a frequent use-case, but it can stand in for any number of confirmed destinations that we want to forward some data to.

You can do the same with Opt 3.